### PR TITLE
feat(github-eventsource): Support multi repos and mitigate race conditions

### LINKED ISSUE #1091: Support for multiple Github repositories to use the same Event
**Is your feature request related to a problem? Please describe.**

For every repository in Github that I want a webhook for - I have to create a new Event since a single Event holds only a single repository. e.g

```
    pull_request:
      owner: example
      repository: demo-api-1  # <- Over here
      webhook:
        endpoint: /npm-pr
        port: "13000"
        method: POST
        url: https://example.com
      events:
        - "pull_request"
      apiToken:
        name: github-access
        key: token
      webhookSecret:
        name: github-access
        key: secret
      active: true
      contentType: json
  ```

**Describe the solution you'd like**

It would be great to be able to have a comma-delimited list of repos in the event instead of having only a single repo per event. e.g solution :

```
    pull_request:
      owner: example
      repository: demo-api-1, demo-api-2    <- Over here
      webhook:
        endpoint: /npm-pr
        port: "13000"
        method: POST
        url: https://example.com
      events:
        - "pull_request"
      apiToken:
        name: github-access
        key: token
      webhookSecret:
        name: github-access
        key: secret
      active: true
      contentType: json
 ```

**Describe alternatives you've considered**

[This](https://github.com/argoproj/argo-events/issues/976#issuecomment-757516121) comment offers a few workarounds, but, none of them are scalable solutions to this issue. 

**Additional context**

This feature is very useful when multiple repos are going to be using the same event.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### LINKED ISSUE #1177: GitHub Webhooks race condition
**Describe the bug**
We have seen multiple times a race condition when GithubEventSoures are evicted, in which the new eventSource receives the hook configuration before it has been deleted when `DeleteHookOnFinish` is enabled.

**To Reproduce**
1. Have a github eventsource with `DeleteHookOnFinish` enabled
2. Evict/delete the pod running
3. Watch for the new pod to start
4. If done fast enough, the `DeleteHookOnFinish` action occurs AFTER the new pod has initilised 

**Expected behavior**
There should be a webhook registered to said repo.

**Environment (please complete the following information):**
 - Kubernetes: `v1.18.16-gke.302`
 - Argo: `v2.12.10`
 - Argo Events: `v1.2.1`

**Additional context**
It is not possible to set `maxUnavailable` / `maxAvailable` on the `Deployment` of the EventSource resource, nor is it possible within the scenario to ensure that a webhook is setup even after initialisation.

One suggestion/Solution would be that a reconcile loop/controller be implemented to ensure that the hooks are configured as expected and re-start the initialisation process if not.

Further context/discussion was raised in Argo Slack at: https://argoproj.slack.com/archives/CAM37QYMS/p1618248491217900

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by whynowy**
1. Github event source support multiple repos, deprecated old spec ( `owner` and `repository`, will still work until v1.6).

Old spec:
```yaml
spec:
  github:
    example:
      owner: argoproj
      repository: argo-events
```

New spec:
```yaml
spec:
  github:
    example:
      repositories:
        - owner: argoproj
          names:
            - argo-events
            - argo-workflows
        - owner: argoproj-labs
          names:
            - abc
            - efg
```
2. Mitigate the issue caused by race conditions, see #1177

Closes #1091 
Closes #1177

Signed-off-by: Derek Wang <whynowy@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 1]
```diff
@@ -651,16 +651,26 @@ type PubSubEventSource struct {
...
+	// Orgnization or user name
+	Owner string `json:"owner,omitempty" protobuf:"bytes,1,opt,name=owner"`
+	// Repository names
+	Names []string `json:"names,omitempty" protobuf:"bytes,2,rep,name=names"`
+}
+
 // GithubEventSource refers to event-source for github related events
 type GithubEventSource struct {
-	// Id is the webhook's id
-	ID int64 `json:"id" protobuf:"varint,1,opt,name=id"`
```
Comment (whynowy) [inline pkg/apis/eventsource/v1alpha1/types.go:657]: `ID` is not used anywhere.

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 2]
```diff
@@ -691,6 +701,25 @@ type GithubEventSource struct {
...
 	Metadata map[string]string `json:"metadata,omitempty" protobuf:"bytes,14,rep,name=metadata"`
+	// Repositories holds the information of repositories, which uses repo owner as the key,
+	// and list of repo names as the value
+	Repositories []OwnedRepositories `json:"repositories,omitempty" protobuf:"bytes,15,rep,name=repositories"`
+}
+
+func (g GithubEventSource) GetOwnedRepositories() []OwnedRepositories {
+	if len(g.Repositories) > 0 {
+		return g.Repositories
+	} else if g.DeprecatedOwner != "" && g.DeprecatedRepository != "" {
```
Comment (alexec) [inline pkg/apis/eventsource/v1alpha1/types.go:712]: this is exactly how I'd do this

File: eventsources/sources/github/validate.go [hunk 3]
```diff
@@ -31,15 +31,13 @@ func validate(githubEventSource *v1alpha1.GithubEventSource) error {
...
-		return fmt.Errorf("repository cannot be empty")
+	if githubEventSource.GetOwnedRepositories() == nil {
+		return fmt.Errorf("no valid repository owner and name found")
 	}
-	if githubEventSource.Owner == "" {
-		return fmt.Errorf("owner cannot be empty")
-	}
-	if githubEventSource.Events == nil || len(githubEventSource.Events) < 1 {
-		return fmt.Errorf("events must be defined")
+	if githubEventSource.APIToken != nil && githubEventSource.Webhook.URL != "" && len(githubEventSource.Events) == 0 {
```
Comment (alexec) [inline eventsources/sources/github/validate.go:37]: minor - you could add githubEventSource.IsValid

File: eventsources/sources/github/types.go [hunk 4]
```diff
@@ -56,8 +56,9 @@ type Router struct {
 	githubEventSource *v1alpha1.GithubEventSource
 	// githubClient is the client to connect to GitHub
 	githubClient *github.Client
-	// hook represents a GitHub (web and service) hook for a repository.
-	hook *github.Hook
+	// hookIDs represents GitHub (web and service) hook ID for repositories.
+	// The map key is the combo of owner and repo name
+	hookIDs map[string]int64
```
Comment (alexec) [inline eventsources/sources/github/types.go:61]: suggestion - I like to say what the key and value are like this:

```golang
hookIDs map[string]int64 // (owner + repo name) -> hook ID
```

File: eventsources/sources/github/start.go [hunk 5]
```diff
@@ -124,182 +125,183 @@ func (router *Router) HandleRoute(writer http.ResponseWriter, request *http.Requ
...
+		r1 := rand.New(s1)
+		time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
+		f()
+
+		go func() {
+			// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
+			// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from github.
+			// This is a workround to mitigate the race conditions.
+			logger.Info("starting github hooks manager daemon")
+			ticker := time.NewTicker(60 * time.Second)
```
Comment (alexec) [inline eventsources/sources/github/start.go:289]: tickers leak references, I prefer to just put a sleep in the default branch

File: eventsources/sources/github/start.go [hunk 6]
```diff
@@ -124,182 +125,183 @@ func (router *Router) HandleRoute(writer http.ResponseWriter, request *http.Requ
...
+			for {
+				select {
+				case <-ctx.Done():
```
Comment (alexec) [inline eventsources/sources/github/start.go:292]: I think we're all getting better using contexts - good stuff

File: eventsources/sources/github/start.go [hunk 7]
```diff
@@ -124,182 +125,183 @@ func (router *Router) HandleRoute(writer http.ResponseWriter, request *http.Requ
...
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
 
-	githubEventSource := &el.GithubEventSource
+		f := func() {
+			for _, r := range githubEventSource.GetOwnedRepositories() {
+				for _, name := range r.Names {
+					hooks, _, err := router.githubClient.Repositories.ListHooks(ctx, r.Owner, name, nil)
+					if err != nil {
+						logger.Errorf("failed to list existing webhooks of %s/%s. err: %+v", r.Owner, name, err)
```
Comment (alexec) [inline eventsources/sources/github/start.go:257]: good clear logging of errors + related info 🥇

File: eventsources/sources/github/start.go [hunk 8]
```diff
@@ -124,182 +125,183 @@ func (router *Router) HandleRoute(writer http.ResponseWriter, request *http.Requ
...
-	if githubEventSource.DeleteHookOnFinish {
-		logger := router.route.Logger.With(
-			"repository", githubEventSource.Repository,
-			"hook-id", *router.hook.ID,
-		)
+		logger.Info("setting up base url for GitHub client...")
+		if githubEventSource.GithubBaseURL != "" {
+			baseURL, err := url.Parse(githubEventSource.GithubBaseURL)
+			if err != nil {
+				return errors.Errorf("failed to parse github base url. err: %s", err)
```
Comment (alexec) [inline eventsources/sources/github/start.go:227]: suggestion - I prefer `fmt.Errorf` its both core Go, and less verbose than `errors.Errorf`

**Review by alexec — APPROVED** [review]
some suggestions

File: eventsources/sources/github/start.go (hunk 5 above)
Comment (whynowy) [reply eventsources/sources/github/start.go:289]: Could you give more detail about that? I know `time.Tick()` "leaks", but never heard `NewTicker` has an issue.

But I did forget to add `defer ticker.Stop()`.

Comment (alexec) [reply eventsources/sources/github/start.go:289]: Well there you go.

File: eventsources/sources/github/validate.go (hunk 3 above)
Comment (whynowy) [reply eventsources/sources/github/validate.go]: Updated.

File: eventsources/sources/github/start.go (hunk 5 above)
Comment (whynowy) [reply eventsources/sources/github/start.go:286]: Fixed all the same issues in the project.

File: eventsources/sources/github/start.go (hunk 8 above)
Comment (whynowy) [reply eventsources/sources/github/start.go]: done.

**Pull request merged**